flink-streaming-java/src/test/java/org/apache/flink/streaming/runtime/tasks/StreamTaskTest.java

```diff
@@ -1864,6 +1866,61 @@ static void processMailTillCheckpointSucceeds(
         testHarness.getTaskStateManager().getWaitForReportLatch().await();
     }
 
+    @Test
+    public void testWaitingForPendingCheckpointsOnFinished() throws Exception {
```

Comment:
Currently, the test passes both with and without waiting for the pending checkpoints.

Moreover I think there is a lot of timing assumptions that need to be met for the test to verify the intended behaviour. I could not think of a good way to fix all of them (would be best to remove the remaining `Thread#sleep()`), but what do you think about such test:

```
    @Test
    public void testWaitingForPendingCheckpointsOnFinished() throws Exception {
        OneShotLatch asyncCheckpointExecuted = new OneShotLatch();
        OneShotLatch canCheckpointBeAcknowledged = new OneShotLatch();
        OneShotLatch invokeCompleted = new OneShotLatch();
        TestCheckpointResponder responder =
                new TestCheckpointResponder() {
                    @Override
                    public void acknowledgeCheckpoint(
                            JobID jobID,
                            ExecutionAttemptID executionAttemptID,
                            long checkpointId,
                            CheckpointMetrics checkpointMetrics,
                            TaskStateSnapshot subtaskState) {

                        try {
                            asyncCheckpointExecuted.trigger();
                            canCheckpointBeAcknowledged.await();
                        } catch (Exception e) {
                            throw new RuntimeException(e);
                        }
                    }
                };

        CompletableFuture<Void> taskClosed =
                CompletableFuture.runAsync(
                        () -> {
                            try (StreamTaskMailboxTestHarness<String> harness =
                                    new StreamTaskMailboxTestHarnessBuilder<>(
                                                    OneInputStreamTask::new,
                                                    BasicTypeInfo.STRING_TYPE_INFO)
                                            .addInput(BasicTypeInfo.STRING_TYPE_INFO, 3)
                                            .modifyStreamConfig(
                                                    config -> config.setCheckpointingEnabled(true))
                                            .setCheckpointResponder(responder)
                                            .setupOperatorChain(new EmptyOperator())
                                            .finishForSingletonOperatorChain(
                                                    StringSerializer.INSTANCE)
                                            .build()) {

                                harness.streamTask
                                        .getCheckpointCoordinator()
                                        .setEnableCheckpointAfterTasksFinished(true);

                                harness.streamTask.triggerCheckpointOnBarrier(
                                        new CheckpointMetaData(1, 101),
                                        CheckpointOptions.forCheckpointWithDefaultLocation(),
                                        new CheckpointMetricsBuilder()
                                                .setBytesProcessedDuringAlignment(0L)
                                                .setAlignmentDurationNanos(0L));

                                harness.waitForTaskCompletion();
                                invokeCompleted.trigger();
                                harness.finishProcessing();
                            } catch (Exception e) {
                                e.printStackTrace();
                            }
                        },
                        Executors.newSingleThreadExecutor());

        asyncCheckpointExecuted.await();
        invokeCompleted.await();
        Thread.sleep(500); // give some potential time for the task to finish before the
        // checkpoint is acknowledged
        assertFalse(taskClosed.isDone());
        canCheckpointBeAcknowledged.trigger();
        taskClosed.get();
    }
```

Reply:
Sorry for not check if the test is work in reverse cases... It seems indeed not easy to accurately test "wait", and I think the suggested implementation is very good. I have changed  the test to this implementation~